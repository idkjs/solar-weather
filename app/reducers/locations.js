import moment from 'moment';
import * as types from '../actions/types.action';

export const initialState = (date) => {
  return {
    locations: [{
      id: 1,
      lat: 0.0,
      lng: 0.0,
      currently: {
        icon: '',
        precipProbability: 0.00,
        precipType: '',
        temperature: 0.00,
        apparentTemperature: 0.00,
        humidity: 0,
        summary: '',
      },
      daily: {
        data: [],
      },
      hourly: {
        data: [],
      },
      alerts: [],
    }],
    timezone: 'America/New_York',
    locationError: null,
    loading: false,
    latestCollectiveUpdate: date || moment().subtract(1, 'day'),
  };
};

export default function locations(state = initialState(), action = {}) {
  switch (action.type) {
    case types.LOCATION_LOADING_OFF:
      return {
        ...state,
        loading: false,
      };
    case types.FETCH_LOCATIONS:
    case types.LOCATION_LOADING:
      return {
        ...state,
        loading: true,
      };
    case types.GET_LOCATIONS:
      return {
        ...state,
        locations: action.locations,
        locationError: null,
        loading: false,
      };
    case types.SET_LOCATION: // eslint-disable-line
      const currentLocation = Object.assign(state.locations[action.index], action.location);
      const locs = state.locations;
      locs[action.index] = currentLocation;
      return {
        ...state,
        locations: locs,
        locationError: null,
        loading: false,
      };
    case types.ADD_LOCATION:
      return {
        ...state,
        locations: [...state.locations.filter(x => x.id !== 1), action.location],
        locationError: null,
        loading: false,
      };
    case types.FETCH_LOCATIONS_SUCCESS:
      return {
        ...state,
        locations: action.locations,
        locationError: null,
        latestCollectiveUpdate: moment(),
        loading: false,
      };
    case types.DELETE_LOCATION: // eslint-disable-line
      return {
        ...state,
        locations: [...state.locations.filter(x => x.id !== action.id)],
        locationError: null,
        loading: false,
      };
    case types.FETCH_LOCATIONS_FAILURE:
    case types.ADD_INDEX_LOCATION_ERROR:
    case types.ADD_LOCATION_ERROR:
    case types.SET_ACTIVE_LOCATION_ERROR:
    case types.UPDATE_ERROR:
      return {
        ...state,
        locationError: action.err,
        loading: false,
      };
    default:
      return state;
  }
}
