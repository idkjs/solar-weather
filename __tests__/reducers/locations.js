import * as types from '../../app/actions/types.action';
import locations, { initialState } from '../../app/reducers/locations';

const loc = {
  id: 1,
  lat: 0.0,
  lng: 0.0,
  currently: {
    icon: '',
    precipProbability: 0.00,
    precipType: '',
    temperature: 0.00,
    apparentTemperature: 0.00,
    humidity: 0,
    summary: '',
  },
  daily: {
    data: [],
  },
  hourly: {
    data: [],
  },
  alerts: [],
};

const date = new Date(1515083977 * 1000);
const state = initialState(date);

describe('locations reducer', () => {
  it('should return the initial state', () => {
    expect(locations(state, {})).toMatchSnapshot();
  });

  it('should handle LOCATION_LOADING', () => {
    expect(
      locations(state, {
        type: types.LOCATION_LOADING,
      })
    ).toMatchSnapshot();
  });

  it('should handle ADD_LOCATION and add a location to state', () => {
    expect(
      locations(state, {
        type: types.ADD_LOCATION,
        location: loc,
      })
    ).toMatchSnapshot();
  });

  it('should handle GET_LOCATIONS and add them to state', () => {
    expect(
      locations(state, {
        type: types.GET_LOCATIONS,
        locations: [loc],
      })
    ).toMatchSnapshot();
  });

  it('should handle UPDATE_ERROR', () => {
    expect(
      locations(state, {
        type: types.UPDATE_ERROR,
        err: 'Test Error',
      })
    ).toMatchSnapshot();
  });
});
