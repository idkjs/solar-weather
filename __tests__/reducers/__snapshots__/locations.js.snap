// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`locations reducer should handle ADD_LOCATION and add a location to state 1`] = `
Object {
  "latestCollectiveUpdate": 2018-01-04T16:39:37.000Z,
  "loading": false,
  "locationError": null,
  "locations": Array [
    Object {
      "alerts": Array [],
      "currently": Object {
        "apparentTemperature": 0,
        "humidity": 0,
        "icon": "",
        "precipProbability": 0,
        "precipType": "",
        "summary": "",
        "temperature": 0,
      },
      "daily": Object {
        "data": Array [],
      },
      "hourly": Object {
        "data": Array [],
      },
      "id": 1,
      "lat": 0,
      "lng": 0,
    },
  ],
  "timezone": "America/New_York",
}
`;

exports[`locations reducer should handle GET_LOCATIONS and add them to state 1`] = `
Object {
  "latestCollectiveUpdate": 2018-01-04T16:39:37.000Z,
  "loading": false,
  "locationError": null,
  "locations": Array [
    Object {
      "alerts": Array [],
      "currently": Object {
        "apparentTemperature": 0,
        "humidity": 0,
        "icon": "",
        "precipProbability": 0,
        "precipType": "",
        "summary": "",
        "temperature": 0,
      },
      "daily": Object {
        "data": Array [],
      },
      "hourly": Object {
        "data": Array [],
      },
      "id": 1,
      "lat": 0,
      "lng": 0,
    },
  ],
  "timezone": "America/New_York",
}
`;

exports[`locations reducer should handle LOCATION_LOADING 1`] = `
Object {
  "latestCollectiveUpdate": 2018-01-04T16:39:37.000Z,
  "loading": true,
  "locationError": null,
  "locations": Array [
    Object {
      "alerts": Array [],
      "currently": Object {
        "apparentTemperature": 0,
        "humidity": 0,
        "icon": "",
        "precipProbability": 0,
        "precipType": "",
        "summary": "",
        "temperature": 0,
      },
      "daily": Object {
        "data": Array [],
      },
      "hourly": Object {
        "data": Array [],
      },
      "id": 1,
      "lat": 0,
      "lng": 0,
    },
  ],
  "timezone": "America/New_York",
}
`;

exports[`locations reducer should handle UPDATE_ERROR 1`] = `
Object {
  "latestCollectiveUpdate": 2018-01-04T16:39:37.000Z,
  "loading": false,
  "locationError": "Test Error",
  "locations": Array [
    Object {
      "alerts": Array [],
      "currently": Object {
        "apparentTemperature": 0,
        "humidity": 0,
        "icon": "",
        "precipProbability": 0,
        "precipType": "",
        "summary": "",
        "temperature": 0,
      },
      "daily": Object {
        "data": Array [],
      },
      "hourly": Object {
        "data": Array [],
      },
      "id": 1,
      "lat": 0,
      "lng": 0,
    },
  ],
  "timezone": "America/New_York",
}
`;

exports[`locations reducer should return the initial state 1`] = `
Object {
  "latestCollectiveUpdate": 2018-01-04T16:39:37.000Z,
  "loading": false,
  "locationError": null,
  "locations": Array [
    Object {
      "alerts": Array [],
      "currently": Object {
        "apparentTemperature": 0,
        "humidity": 0,
        "icon": "",
        "precipProbability": 0,
        "precipType": "",
        "summary": "",
        "temperature": 0,
      },
      "daily": Object {
        "data": Array [],
      },
      "hourly": Object {
        "data": Array [],
      },
      "id": 1,
      "lat": 0,
      "lng": 0,
    },
  ],
  "timezone": "America/New_York",
}
`;
